File: src/widgets/datetime.zig
const std = @import("std");
const themes = @import("../themes/registry.zig");

const TimeFormat = @import("../cli/args.zig").TimeFormat;

// Cache the formatted time string within a single process invocation.
// Note: tmux #() spawns a fresh process per refresh, so this only helps
// if the same binary is reused within the same minute (e.g., multiple
// widgets invoked in one batch). The primary benefit is avoiding the
// `date` fork when the binary handles multiple datetime calls.
var last_minute_of_day: u16 = 0xFFFF;
var cached_time: [32]u8 = undefined;
var cached_len: usize = 0;

pub fn run(allocator: std.mem.Allocator, theme_name: []const u8, time_format: TimeFormat, transparent: bool, io: std.Io, environ_map: *std.process.Environ.Map, writer: *std.Io.Writer) !void {
    const theme = (themes.byName(allocator, io, environ_map, theme_name) orelse themes.hard).withTransparentBackground(transparent);

    const separator = "▒";
    const time_icon = "󰥔";

    var time_str: []const u8 = "";
    if (time_format != .hide) {
        const now = std.Io.Timestamp.now(io, .real);
        const total_seconds: u64 = @intCast(@divFloor(now.nanoseconds, std.time.ns_per_s));
        const current_minute: u16 = @intCast((total_seconds / 60) % 1440);

        if (current_minute == last_minute_of_day) {
            time_str = cached_time[0..cached_len];
        } else {
            const fmt = switch (time_format) {
                .H12 => "+%I:%M %p ",
                .H24 => "+%H:%M ",
                .hide => unreachable,
            };

            var buf: [4096]u8 = undefined;
            var fba = std.heap.FixedBufferAllocator.init(&buf);
            const fba_allocator = fba.allocator();

            const result = std.process.run(fba_allocator, io, .{
                .argv = &.{ "date", fmt },
            }) catch null;

            if (result) |r| {
                if (r.term == .exited and r.term.exited == 0) {
                    const trimmed = std.mem.trim(u8, r.stdout, " \n\r\t");
                    if (trimmed.len <= cached_time.len) {
                        @memcpy(cached_time[0..trimmed.len], trimmed);
                        cached_len = trimmed.len;
                        last_minute_of_day = current_minute;
                        time_str = cached_time[0..cached_len];
                    } else {
                        time_str = trimmed;
                    }
                }
            } else {
                // Fallback to native UTC formatting when `date` is unavailable
                const epoch_secs = std.time.epoch.EpochSeconds{ .secs = total_seconds };
                const day_secs = epoch_secs.getDaySeconds();
                const hour = day_secs.getHoursIntoDay();
                const minute = day_secs.getMinutesIntoHour();

                const formatted = if (time_format == .H12)
                    format12h(&cached_time, hour, minute)
                else
                    format24h(&cached_time, hour, minute);
                cached_len = formatted.len;
                last_minute_of_day = current_minute;
                time_str = cached_time[0..cached_len];
            }
        }
    }

    try writer.print("#[fg={s},bg={s}]{s} #[fg={s}]{s} {s}", .{
        theme.accent,
        theme.surface_alt,
        separator,
        theme.emphasis,
        time_icon,
        time_str,
    });
}

fn format24h(buf: *[32]u8, hour: u5, minute: u6) []const u8 {
    return std.fmt.bufPrint(buf[0..], "{d:0>2}:{d:0>2} ", .{ hour, minute }) catch unreachable;
}

fn format12h(buf: *[32]u8, hour: u5, minute: u6) []const u8 {
    const is_pm = hour >= 12;
    const display_hour = if (hour == 0) 12 else if (hour > 12) hour - 12 else hour;
    const suffix = if (is_pm) "PM" else "AM";
    return std.fmt.bufPrint(buf[0..], "{d:0>2}:{d:0>2} {s} ", .{ display_hour, minute, suffix }) catch unreachable;
}

test "format24h produces HH:MM " {
    var buf: [32]u8 = undefined;
    try std.testing.expectEqualStrings("09:05 ", format24h(&buf, 9, 5));
    try std.testing.expectEqualStrings("23:59 ", format24h(&buf, 23, 59));
    try std.testing.expectEqualStrings("00:00 ", format24h(&buf, 0, 0));
}

test "format12h produces HH:MM AM/PM " {
    var buf: [32]u8 = undefined;
    try std.testing.expectEqualStrings("12:00 AM ", format12h(&buf, 0, 0));
    try std.testing.expectEqualStrings("12:30 PM ", format12h(&buf, 12, 30));
    try std.testing.expectEqualStrings("03:45 PM ", format12h(&buf, 15, 45));
    try std.testing.expectEqualStrings("11:59 PM ", format12h(&buf, 23, 59));
}

File: src/widgets/datetime.rs
use crate::cli::args::TimeFormat;
use crate::core::Theme;
use crate::tmux_renderer;

fn format_24h(hour: u32, minute: u32) -> String {
    format!("{hour:02}:{minute:02} ")
}

fn format_12h(hour: u32, minute: u32) -> String {
    let is_pm = hour >= 12;
    let display_hour = match hour {
        0 => 12,
        13..=23 => hour - 12,
        _ => hour,
    };
    let suffix = if is_pm { "PM" } else { "AM" };
    format!("{display_hour:02}:{minute:02} {suffix} ")
}

fn local_time() -> time::Time {
    time::OffsetDateTime::now_local()
        .unwrap_or_else(|_| time::OffsetDateTime::now_utc())
        .time()
}

fn format_time(time_format: TimeFormat, time: time::Time) -> String {
    match time_format {
        TimeFormat::H12 => format_12h(u32::from(time.hour()), u32::from(time.minute())),
        TimeFormat::H24 => format_24h(u32::from(time.hour()), u32::from(time.minute())),
        TimeFormat::Hide => String::new(),
    }
}

/// Render the datetime widget with a configurable 12H/24H/hide format.
pub fn run(theme: Theme, time_format: TimeFormat) -> String {
    let time_str = format_time(time_format, local_time());
    if time_str.is_empty() {
        return String::new();
    }

    let separator = "▒";
    let time_icon = "󰥔";

    format!(
        "#[fg={},bg={}]{} #[fg={}]{} {}",
        tmux_renderer::color_hex_string(theme.accent),
        tmux_renderer::color_hex_string(theme.surface_alt),
        separator,
        tmux_renderer::color_hex_string(theme.emphasis),
        time_icon,
        time_str,
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn format_24h_produces_hh_colon_mm_space() {
        assert_eq!("09:05 ", format_24h(9, 5));
        assert_eq!("23:59 ", format_24h(23, 59));
        assert_eq!("00:00 ", format_24h(0, 0));
    }

    #[test]
    fn format_12h_produces_hh_colon_mm_am_pm_space() {
        assert_eq!("12:00 AM ", format_12h(0, 0));
        assert_eq!("12:30 PM ", format_12h(12, 30));
        assert_eq!("03:45 PM ", format_12h(15, 45));
        assert_eq!("11:59 PM ", format_12h(23, 59));
    }

    #[test]
    fn format_time_hides_time_text() {
        let time = time::Time::from_hms(9, 5, 0).unwrap();
        assert_eq!("", format_time(TimeFormat::Hide, time));
    }
}
